File: src/ee/unity_ads/private/UnityInterstitialAd.cpp
//
//  UnityInterstitialAd.cpp
//  ee_x
//
//  Created by Zinge on 10/10/17.
//
//

#include "ee/unity_ads/private/UnityInterstitialAd.hpp"

#include <cassert>

#include <ee/ads/internal/IAsyncHelper.hpp>
#include <ee/core/Logger.hpp>
#include <ee/core/Utils.hpp>
#include <ee/coroutine/Task.hpp>

#include "ee/unity_ads/UnityAdsBridge.hpp"

namespace ee {
namespace unity_ads {
using Self = InterstitialAd;

Self::InterstitialAd(const Logger& logger,
                     const std::shared_ptr<ads::IAsyncHelper<bool>>& displayer,
                     Bridge* plugin, const std::string& adId)
    : logger_(logger)
    , displayer_(displayer)
    , plugin_(plugin)
    , adId_(adId) {
    logger_.debug(__PRETTY_FUNCTION__);
}

Self::~InterstitialAd() {}

void Self::destroy() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    plugin_->destroyInterstitialAd(adId_);
}

bool Self::isLoaded() const {
    return plugin_->hasRewardedAd(adId_);
}

Task<bool> Self::load() {
    // No effect.
    co_return false;
}

Task<bool> Self::show() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    auto result = co_await displayer_->process(
        [this] { //
            plugin_->showRewardedAd(adId_);
        },
        [](bool result) {
            // OK.
        });
    co_return result;
}

void Self::onLoaded() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    dispatchEvent([&](auto&& observer) {
        if (observer.onLoaded) {
            observer.onLoaded();
        }
    });
}

void Self::onFailedToShow(const std::string& message) {
    logger_.debug("%s: message = %s displaying = %s", __PRETTY_FUNCTION__,
                  message.c_str(),
                  core::toString(displayer_->isProcessing()).c_str());
    if (displayer_->isProcessing()) {
        displayer_->resolve(false);
    } else {
        assert(false);
    }
}

void Self::onClosed() {
    logger_.debug("%s: displaying = %s", __PRETTY_FUNCTION__,
                  core::toString(displayer_->isProcessing()).c_str());
    if (displayer_->isProcessing()) {
        displayer_->resolve(true);
    } else {
        assert(false);
    }
}
} // namespace unity_ads
} // namespace ee

File: unity/Runtime/UnityAds/Internal/UnityInterstitialAd.cs
using System.Threading.Tasks;

using UnityEngine.Assertions;

namespace EE.Internal {
    internal class UnityInterstitialAd
        : ObserverManager<IInterstitialAdObserver>, IInterstitialAd {
        private readonly IAsyncHelper<bool> _displayer;
        private readonly UnityAds _plugin;
        private readonly string _adId;

        public UnityInterstitialAd(
            IAsyncHelper<bool> displayer, UnityAds plugin, string adId) {
            _displayer = displayer;
            _plugin = plugin;
            _adId = adId;
        }

        public void Destroy() {
            _plugin.DestroyInterstitialAd(_adId);
        }

        public bool IsLoaded => _plugin.HasRewardedAd(_adId);

        public Task<bool> Load() {
            // No effect.
            return Task.FromResult(false);
        }

        public Task<bool> Show() {
            return _displayer.Process(
                () => _plugin.ShowRewardedAd(_adId),
                result => {
                    // OK.
                });
        }

        internal void OnLoaded() {
            DispatchEvent(observer => observer?.OnLoaded());
        }

        internal void OnFailedToShow(string message) {
            if (_displayer.IsProcessing) {
                _displayer.Resolve(false);
            } else {
                Assert.IsTrue(false);
            }
        }

        internal void OnClosed() {
            if (_displayer.IsProcessing) {
                _displayer.Resolve(true);
            } else {
                Assert.IsTrue(false);
            }
        }
    }
}